src/main/java/supersymmetry/api/capability/impl/BiomeRecipeLogic.java

```diff
@@ -0,0 +1,47 @@
+package supersymmetry.api.capability.impl;
+
+import gregtech.api.capability.impl.MultiblockRecipeLogic;
+import gregtech.api.metatileentity.multiblock.RecipeMapMultiblockController;
+import gregtech.api.recipes.Recipe;
+import org.jetbrains.annotations.NotNull;
+import supersymmetry.api.recipes.properties.BiomeProperty;
+
+/**
+ * A custom recipeLogic class, for adding our check for biomes
+ * This can be moved out to a stand-alone class.
+ * But generally speaking if you do not plan to re-use this, making it an inner class should be fine.
+ * CEu itself has many such cases.
+ */
+public class BiomeRecipeLogic extends MultiblockRecipeLogic {
+
+    public BiomeRecipeLogic(RecipeMapMultiblockController tileEntity) {
+        super(tileEntity, true);
+    }
+
+
+    /**
+     * Overriding this to add our own custom checks
+     * Don't forget super calls
+     */
+    @Override
+    public boolean checkRecipe(@NotNull Recipe recipe) {
+        return checkHeightRequirement() && checkBiomeRequirement(recipe) && super.checkRecipe(recipe);
+    }
+
+    public boolean checkHeightRequirement() {
+        return getMetaTileEntity().getPos().getY() == 64;
+    }
+    /**
+     * This is a method for biome checking
+     */
+    public boolean checkBiomeRequirement(@NotNull Recipe recipe) {
+        if (!recipe.hasProperty(BiomeProperty.getInstance())) return true;
+        return recipe.getProperty(BiomeProperty.getInstance(), BiomeProperty.BiomePropertyList.EMPTY_LIST)
+                .checkBiome(getMetaTileEntity().getWorld().getBiome(getMetaTileEntity().getPos()));
```

Comment:
btw ig you should check the biome at the position of the pipe casing instead of at the controller

Reply:
can you review this section rn?